### PR TITLE
machinectl: add verb edit and cat to operate on .nspawn files

### LINKED ISSUE #26246: machinectl edit
### Component

machinectl

### Is your feature request related to a problem? Please describe

Simply access to `/etc/systemd/nspawn/machine.nspawn` editor will be very useful.
More information in `machinectl show machine` would be useful too...

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

_No response_

### PR DESCRIPTION
**Pull request by YHNdnzj**
Closes #26246 

~~This requires fixes from #26756, so it's marked as draft at the moment.~~

### CODE REVIEW

**Comment by yuwata** [comment]
@YHNdnzj I guess now you have superpower, hence feel free to merge #26756 if you'd like.

**Comment by YHNdnzj** [comment]
> @YHNdnzj I guess now you have superpower, hence feel free to merge #26756 if you'd like.

I feel a bit unsure if I shall merge it as I'm the only reviewer. But I guess it's fine for this one :)

File: src/machine/machinectl.c [hunk 1]
```diff
@@ -2443,7 +2587,9 @@ static int help(int argc, char *argv[], void *userdata) {
                "  kill NAME...                Send signal to processes of a VM/container\n"
                "  copy-to NAME PATH [PATH]    Copy files from the host to a container\n"
                "  copy-from NAME PATH [PATH]  Copy files from a container to the host\n"
-               "  bind NAME PATH [PATH]       Bind mount a path from the host into a container\n\n"
+               "  bind NAME PATH [PATH]       Bind mount a path from the host into a container\n"
+               "  edit NAME...                Edit settings of one or more VMs/containers\n\n"
```
Comment (yuwata) [inline src/machine/machinectl.c]: doubled `\n` at the end.
I'd like to move one `\n` to the line `Image Commands`, then each verb line has one `\n` at the end.

File: man/machinectl.xml [hunk 2]
```diff
@@ -330,6 +330,24 @@
...
+        <para>Normally, if the settings file of the given machine can't be found, <command>edit</command> aborts
+        the operation and exits with failure. When used with <option>--new</option>, it will instead create
+        a new settings file from scratch.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><command>cat</command> <replaceable>NAME</replaceable></term>
+
+        <listitem><para>Cat the settings file of the specified machines. For the detailed format, see <citerefentry
+        project='man-pages'><refentrytitle>systemd.nspawn</refentrytitle><manvolnum>5</manvolnum></citerefentry>.</para></listitem>
```
Comment (yuwata) [inline man/machinectl.xml]: `For the detailed format...` ?? What format do you want to explain here?

File: src/machine/machinectl.c [hunk 3]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+        _cleanup_free_ char *file = NULL, *path = NULL;
+
+        assert(machine);
+        assert(ret_path);
+
+        file = strjoin(machine, ".nspawn");
+        if (!file)
+                return log_oom();
+
+        FOREACH_STRING(i, "/etc/systemd/nspawn", "/run/systemd/nspawn") {
```
Comment (yuwata) [inline src/machine/machinectl.c]: Please also append `"/var/lib/machines"` here.

I think, `ret_is_var` can be dropped as the caller can be easily check it by using `path_startswith()`.

File: src/machine/machinectl.c [hunk 4]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                        return log_oom();
+
+                if (access(path, F_OK) >= 0) {
+                        *ret_path = TAKE_PTR(path);
+                        if (ret_is_var)
+                                *ret_is_var = false;
+                        return 0;
+                }
+
+                path = mfree(path);
```
Comment (yuwata) [inline src/machine/machinectl.c]: Please move the declaration of `path` into the loop, and drop `mfree()` here.

File: src/machine/machinectl.c [hunk 5]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+        if (!IN_SET(arg_transport, BUS_TRANSPORT_LOCAL, BUS_TRANSPORT_MACHINE))
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "Edit is only supported on local machines.");
+
+        r = mac_selinux_init();
+        if (r < 0)
+                return r;
+
+        if (!arg_new) {
+                STRV_FOREACH(machine, argv + 1) {
```
Comment (yuwata) [inline src/machine/machinectl.c]: Maybe, better to introduce `char **files = strv_skip(argv, 1);`

File: src/machine/machinectl.c [hunk 6]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+static int edit_settings(int argc, char *argv[], void *userdata) {
+        _cleanup_(edit_file_context_done) EditFileContext context = {
+                .remove_parent = false,
+        };
+        int r;
+
+        if (!on_tty())
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Cannot edit machine settings if not on a tty.");
+
+        if (!IN_SET(arg_transport, BUS_TRANSPORT_LOCAL, BUS_TRANSPORT_MACHINE))
```
Comment (yuwata) [inline src/machine/machinectl.c]: How does `BUS_TRANSPORT_MACHINE` work ??

File: src/machine/machinectl.c [hunk 7]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                }
+        }
+
+        return do_edit_files_and_install(&context);
+}
+
+static int cat_settings(int argc, char *argv[], void *userdata) {
+        int r = 0;
+
+        if (!IN_SET(arg_transport, BUS_TRANSPORT_LOCAL, BUS_TRANSPORT_MACHINE))
```
Comment (yuwata) [inline src/machine/machinectl.c]: Also here. I guess only local works for these commands.

File: src/machine/machinectl.c [hunk 8]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "Cat is only supported on local machines.");
+
+        pager_open(arg_pager_flags);
+
+        STRV_FOREACH(machine, argv + 1) {
```
Comment (yuwata) [inline src/machine/machinectl.c]: `strv_skip(argv, 1)`

File: src/machine/machinectl.c [hunk 9]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                q = get_settings_path(*machine, &path, NULL);
+                if (q == -ENOENT) {
+                        log_error_errno(q, "No settings file found for machine '%s'.", *machine);
+                        r = r < 0 ? r : q;
+                        continue;
+                }
+                if (q < 0)
+                        return r < 0 ? r : q;
+
+                q = cat_files(path, 0, 0);
```
Comment (yuwata) [inline src/machine/machinectl.c]: `cat_files(path, /* dropins = */ NULL, /* flags = */ 0)`

File: src/machine/machinectl.c [hunk 10]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                int q;
+
+                q = get_settings_path(*machine, &path, NULL);
+                if (q == -ENOENT) {
+                        log_error_errno(q, "No settings file found for machine '%s'.", *machine);
+                        r = r < 0 ? r : q;
+                        continue;
+                }
+                if (q < 0)
+                        return r < 0 ? r : q;
```
Comment (yuwata) [inline src/machine/machinectl.c]: Please log about the failure.

File: src/machine/machinectl.c [hunk 11]
```diff
@@ -1412,6 +1414,149 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+
+                q = cat_files(path, 0, 0);
+                if (q < 0)
+                        return r < 0 ? r : q;
```
Comment (yuwata) [inline src/machine/machinectl.c:1567]: Also here.

File: man/machinectl.xml [hunk 12]
```diff
@@ -330,6 +330,24 @@
...
       </varlistentry>
+
+      <varlistentry>
+        <term><command>edit</command> <replaceable>NAME</replaceable></term>
+
+        <listitem><para>Edit the settings file of the specified machines. For the detailed format, see <citerefentry
+        project='man-pages'><refentrytitle>systemd.nspawn</refentrytitle><manvolnum>5</manvolnum></citerefentry>.</para>
+
+        <para>Normally, if the settings file of the given machine can't be found, <command>edit</command> aborts
+        the operation and exits with failure. When used with <option>--new</option>, it will instead create
```
Comment (yuwata) [inline man/machinectl.xml]: Hmm... `--new` is necessary? Why not gracefully create a new file?

**Review by yuwata — CHANGES_REQUESTED** [review]
Please add a brief test for the new verbs.

File: src/machine/machinectl.c (hunk 11 above)
Comment (YHNdnzj) [reply src/machine/machinectl.c:1567]: This and the one above have been logged by the callee I believe 🤔

File: src/machine/machinectl.c [hunk 13]
```diff
@@ -1412,6 +1413,119 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+        _cleanup_free_ char *file_alloc = NULL;
+        const char *file;
+
+        assert(name);
+        assert(ret_path);
+
+        if (!endswith(name, ".nspawn")) {
+                file_alloc = strjoin(name, ".nspawn");
+                if (!file_alloc)
+                        return log_oom();
```
Comment (yuwata) [inline src/machine/machinectl.c]: Please make each function _always_ log failure or silently (except for log_debug) return negative errno.

This is a kind of library like function, hence `return -ENOMEM;` should be sufficient.

File: src/machine/machinectl.c [hunk 14]
```diff
@@ -1412,6 +1413,119 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                file = file_alloc;
+        } else
+                file = name;
+
+        FOREACH_STRING(i, "/etc/systemd/nspawn", "/run/systemd/nspawn", "/var/lib/machines") {
+                _cleanup_free_ char *path = NULL;
+
+                path = path_join(i, file);
+                if (!path)
+                        return log_oom();
```
Comment (yuwata) [inline src/machine/machinectl.c]: `return -ENOMEM;`

File: src/machine/machinectl.c [hunk 15]
```diff
@@ -1412,6 +1413,119 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+
+                file = strjoin(*machine, ".nspawn");
+                if (!file)
+                        return log_oom();
+
+                r = get_settings_path(file, &path, NULL);
+                if (r < 0) {
+                        log_debug_errno(r,
+                                        "No existing settings file for machine '%s' found, creating.",
+                                        *machine);
```
Comment (yuwata) [inline src/machine/machinectl.c]: This is only when `-ENOENT`. And on `-ENOMEM`, we should not create new file:
```c
if (r == -ENOENT) {
        log_debug_errno(...);
        ...;
} else if (r < 0)
        return log_error_errno(r, "Failed to ...");
```

`"creating."` -> `"creating a new file."` is better?

File: src/machine/machinectl.c (hunk 11 above)
Comment (yuwata) [reply src/machine/machinectl.c:1567]: Indeed. `cat_files()` logs its failure.

File: src/machine/machinectl.c [hunk 16]
```diff
@@ -1412,6 +1413,119 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+        r = mac_selinux_init();
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH(machine, strv_skip(argv, 1)) {
+                _cleanup_free_ char *file = NULL, *path = NULL;
+
+                file = strjoin(*machine, ".nspawn");
+                if (!file)
+                        return log_oom();
```
Comment (yuwata) [inline src/machine/machinectl.c]: The same thing is also done in `get_settings_path()`.

**Review by yuwata — CHANGES_REQUESTED** [review]
BTW, I'd like to make these command also take full path.

File: src/machine/machinectl.c [hunk 17]
```diff
@@ -1412,6 +1413,151 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                                       "Edit is only supported on the host machine.");
+
+        r = mac_selinux_init();
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH(name, strv_skip(argv, 1)) {
+                _cleanup_free_ char *path = NULL;
+
+                if (is_path(*name)) {
```
Comment (yuwata) [inline src/machine/machinectl.c]: `path_is_absolute()` and `path_is_safe()` may be better?
```
if (path_is_absolute(*name)) {
        if (!path_is_safe(*name))
                return log_error_errno(...);

        r = edit_files_add(...);
        if (r < 0)
                return r;
        continue;
}

if (!filename_is_valid(*name))
        return log_error_errno(...);

...
```

File: src/machine/machinectl.c [hunk 18]
```diff
@@ -1412,6 +1413,151 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                        if (path_is_valid(*name)) {
+                                r = edit_files_add(&context, *name, NULL, NULL);
+                                continue;
```
Comment (yuwata) [inline src/machine/machinectl.c]: Missing error handling.

File: src/machine/machinectl.c [hunk 19]
```diff
@@ -1412,6 +1413,151 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                        } else
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
```
Comment (yuwata) [inline src/machine/machinectl.c]: Like suggested in the above, I'd like to locate the error handling earlier.

File: src/machine/machinectl.c [hunk 20]
```diff
@@ -1412,6 +1413,151 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+
+                r = get_settings_path(*name, &path);
+                if (r == -ENOENT) {
+                        _cleanup_free_ char *file = NULL;
+
+                        log_debug_errno(r,
+                                        "No existing settings file for machine '%s' found, creating a new file.",
+                                        *name);
+
+                        file = strjoin(*name, ".nspawn");
```
Comment (yuwata) [inline src/machine/machinectl.c]: `*name` may be already has `.nspawn` suffix.

File: src/machine/machinectl.c [hunk 21]
```diff
@@ -1412,6 +1413,151 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                                return r;
+                        continue;
+                }
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get the path of the settings file: %m");
+
+                if (path_startswith(path, "/var/lib/machines")) {
+                        _cleanup_free_ char *file = NULL, *new_path = NULL;
+
+                        file = strjoin(*name, ".nspawn");
```
Comment (yuwata) [inline src/machine/machinectl.c]: Also here.

File: test/units/testsuite-80.sh [hunk 22]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
```
Comment (yuwata) [inline test/units/testsuite-80.sh]: Please merge the test with TEST-74, i.e. rename `testsuite-80.sh` -> `testsuite-74.machinectl.sh`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: man/machinectl.xml [hunk 23]
```diff
@@ -330,6 +330,21 @@
         group ID is preserved for the copy, otherwise all files and directories in the copy will be owned by the root
         user and group (UID/GID 0).</para></listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><command>edit</command> <replaceable>NAME</replaceable></term>
```
Comment (yuwata) [inline man/machinectl.xml]: Please mention that can take full path.
`NAME` -> `NAME|FILENAME`

File: man/machinectl.xml [hunk 24]
```diff
@@ -330,6 +330,21 @@
...
+
+        <listitem><para>Edit the settings file of the specified machines. For the format of the settings file, refer to <citerefentry
+        project='man-pages'><refentrytitle>systemd.nspawn</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
+        If a existing settings file of the given machine can't be found, <command>edit</command> automatically
+        create a new settings file from scratch under <filename>/etc/</filename></para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><command>cat</command> <replaceable>NAME</replaceable></term>
```
Comment (yuwata) [inline man/machinectl.xml]: Also here.

File: man/machinectl.xml [hunk 25]
```diff
@@ -330,6 +330,21 @@
...
+        <listitem><para>Edit the settings file of the specified machines. For the format of the settings file, refer to <citerefentry
+        project='man-pages'><refentrytitle>systemd.nspawn</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
+        If a existing settings file of the given machine can't be found, <command>edit</command> automatically
+        create a new settings file from scratch under <filename>/etc/</filename></para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><command>cat</command> <replaceable>NAME|FILE</replaceable></term>
+
+        <listitem><para>Cat the settings file of the specified machines.</para></listitem>
```
Comment (yuwata) [inline man/machinectl.xml]: `Cat` -> `Show`

File: src/machine/machinectl.c [hunk 26]
```diff
@@ -1412,6 +1413,165 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+                        continue;
+                }
+
+                r = normalize_nspawn_filename(*name, &file);
+                if (r < 0)
+                        return r;
+
+                r = get_settings_path(file, &path);
+                if (r == -ENOENT) {
+                        log_debug_errno(r,
```
Comment (yuwata) [inline src/machine/machinectl.c]: Not necessary to include the error cause, please use `log_debug()`.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by yuwata** [comment]
```
../src/machine/machinectl.c: In function ‘cat_settings’:
../src/machine/machinectl.c:1544:29: error: implicit declaration of function ‘cat_file’; did you mean ‘cat_files’? [-Werror=implicit-function-declaration]
 1544 |                         q = cat_file(*name, /* newline = */ false);
      |                             ^~~~~~~~
      |                             cat_files
../src/machine/machinectl.c:1544:29: error: nested extern declaration of ‘cat_file’ [-Werror=nested-externs]
cc1: all warnings being treated as errors
```

File: src/machine/machinectl.c [hunk 27]
```diff
@@ -1412,6 +1413,163 @@ static int shell_machine(int argc, char *argv[], void *userdata) {
...
+
+        *ret_file = TAKE_PTR(file);
+        return 0;
+}
+
+static int get_settings_path(const char *name, char **ret_path) {
+        assert(name);
+        assert(ret_path);
+
+        FOREACH_STRING(i, "/etc/systemd/nspawn", "/run/systemd/nspawn", "/var/lib/machines") {
```
Comment (poettering) [inline src/machine/machinectl.c:1440]: hmm, i am tempted to say we should always edit the file in /etc, as that is where user config belongs.

Comment (poettering) [reply src/machine/machinectl.c:1440]: ah, don't mind me

File: src/machine/machinectl.c [hunk 28]
```diff
@@ -2443,8 +2601,10 @@ static int help(int argc, char *argv[], void *userdata) {
                "  kill NAME...                Send signal to processes of a VM/container\n"
                "  copy-to NAME PATH [PATH]    Copy files from the host to a container\n"
                "  copy-from NAME PATH [PATH]  Copy files from a container to the host\n"
-               "  bind NAME PATH [PATH]       Bind mount a path from the host into a container\n\n"
-               "Image Commands:\n"
+               "  bind NAME PATH [PATH]       Bind mount a path from the host into a container\n"
+               "  edit NAME|FILE...           Edit settings of one or more VMs/containers\n"
+               "  cat NAME|FILE...            Show settings of one or more VMs/containers\n"
```
Comment (poettering) [inline src/machine/machinectl.c]: these commands do not operate on "live" containers, but on images. hence this should be in the image commands section below

**Comment by poettering** [comment]
lgtm, thanks for working on this

**Review by yuwata — APPROVED** [review]

**Pull request merged**